### PR TITLE
TelescopeControl Plugin Fixes

### LINKED ISSUE #2713: cannot add new mount to telescope-extension
### Expected Behaviour
Adding a new telescope (mount) to stellarium telescope plugin can be finished by clicking "OK" (previously made configuration is stored - telescope configuration window is closed). 

### Actual Behaviour
"OK" can be clicked but has no function - config is not stored and dialog can only be exited by cancel (or "X" in upper corner)

### Steps to reproduce
Fresh installation of: Windows, .net 3.5, ASCOM 6.6, iOptron / or EQMOD for Skywatcher (both same result regarding "actual behavior"), stellarium 1.22.3

actual steps:
open configuration window [F2] -> goto TAB "Plugins" -> choose "Telescope control" -> tic load at startup and restart stellarium
open configuration window [F2] -> goto TAB "Plugins" -> choose "Telescope control" -> configure -> add new telescope (list of telescopes is empty = new installation) -> choose your config (simulation or ASCOM had been tried) -> press "OK" (but this has no function and no effect - window is still open, config not saved)

Same steps tested after deinstallation of stellarium 1.22.3 win64 and reinstall of version 0.22.2 win64 and everything works as expected - new telescope (mount) is recognized and can be slewed with stellarium 0.22.2

### System
* Stellarium version: 1.22.3 win64
* Operating system: Windows 10

### PR DESCRIPTION
**Pull request by gzotti**
### Description

This is an attempt to start modernizing the TelescopeControl Plugin

Some parts of the code are not very "Qt"e or even object-oriented C++. Functions dispersed between objects, etc, some namespace definitions where definitions should just go into a base class, etc.

This branch so far 
- provides a few syntactic improvements 
- encapsulates some weirdly dispersed functions into the classes where they are actually used exclusively
- collects enums into the plugins's main class and declares them in Qt style
- introduces StelProperties also to this plugin
- fixes #2713
- fixes an iterator loop
- adds abort function to the TelescopeDummy
- adds "abort slew" button to the GUI

In addition, I just found (qcompilerdetection.h) that two Qt defines, Q_NULLPTR and Q_DECL_OVERRIDE, which I had previously enforced as being more visible than their standard terms (nullptr, override), should actually be ported away. While we are at it, we can start on this with this plugin. 

I am not too familiar with this plugin. It seems to refer to functionality like external telescope server programs available around 2007 or so. If they are no longer available, this may be cleaned up. If they are still available, their interface and existence should be documented. There are some FIXME comments interspersed which may have been left ten or more years ago. Can we decide upon them? 

Fixes #2713  (issue)

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update

### How Has This Been Tested?

**Test Configuration**:
* Operating system: Windows 10
* Graphics Card: (irrelevant)
* Qt 6.4.0, Qt5.12.8
* I cannot test physical telescopes. Compilation and telescope simulation dummy work. 

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by alex-w** [comment]
When INDI/INDIGO connection type is used the button for get the list of device is stay "pressed" when connection to the INDI/INDIGO server is fails

**Comment by gzotti** [comment]
There are a few reports about problems in the TC plugin which we should want to solve in this PR. 
Has this been brought in during this work, or is this also in V1.0?

So far the updates were quite mechanically and should not have caused new issues. :-P This plugin is usually disabled and not even built on my system, so I am not too familiar with it. My last use was around 2017...  I have neither ASCOM nor INDI installed on my systems, so I cannot test it properly. I only have the telescope dummy. 

I have seen a few problematic constructions in here, and lots of TODOs and FIXMEs marked since about 2010 (BM, FC!). If you operate instruments, maybe you can find out what actually the problems are. Apparently having an ASCOM device connected at start is a problem (#2717)? No idea about INDI, sorry. Or somebody else please?

**Review by alex-w — APPROVED** [review]

**Comment by gzotti** [comment]
I actually don't want fast approval but clean up the plugin... Just that I don't know its (pre-)history.  Will continue a bit. If something is broken and we see it, we should fix it.

**Comment by alex-w** [comment]
> Has this been brought in during this work, or is this also in V1.0?

It is in v1.0 too... I found this issue when tested your branch, but forget run INDIGO server.

File: plugins/TelescopeControl/src/ASCOM/TelescopeClientASCOM.cpp [hunk 1]
```diff
@@ -99,7 +80,7 @@ void TelescopeClientASCOM::performCommunication()
 			  mDoesRefraction ? StelCore::RefractionOff : StelCore::RefractionOn);
 		}
 
-		if (!(areSimilar(mLastCoord.RA, currentCoords.RA) && areSimilar(mLastCoord.DEC, currentCoords.DEC)))
```
Comment (ShahriarNasseri) [inline plugins/TelescopeControl/src/ASCOM/TelescopeClientASCOM.cpp:102]: check the resolution and vector analysis

File: plugins/TelescopeControl/src/Lx200/TelescopeClientDirectLx200.hpp [hunk 2]
```diff
@@ -89,7 +90,7 @@ class TelescopeClientDirectLx200 : public TelescopeClient, public Server
 		return interpolatedPosition.isKnown();
 	}
 
-	Equinox equinox;
```
Comment (ShahriarNasseri) [inline plugins/TelescopeControl/src/Lx200/TelescopeClientDirectLx200.hpp:92]: schemmatic of the vectors is the same

File: plugins/TelescopeControl/src/ASCOM/TelescopeClientASCOM.cpp [hunk 3]
```diff
@@ -28,26 +28,7 @@
 #include "StelUtils.hpp"
 #include "StelTranslator.hpp"
 
-bool useJNow(ASCOMDevice::ASCOMEquatorialCoordinateType coordinateType, bool mAscomUseDeviceEqCoordType, Equinox mEquinox)
```
Comment (ShahriarNasseri) [inline plugins/TelescopeControl/src/ASCOM/TelescopeClientASCOM.cpp:31]: the system file format is the act of the mirrors

Comment (gzotti) [reply plugins/TelescopeControl/src/ASCOM/TelescopeClientASCOM.cpp:31]: I have move this into a static method of the class that uses it (encapsulation). Can you elaborate what problem you see?

File: plugins/TelescopeControl/src/Lx200/TelescopeClientDirectLx200.hpp (hunk 2 above)
Comment (gzotti) [reply plugins/TelescopeControl/src/Lx200/TelescopeClientDirectLx200.hpp:92]: schematic of which vectors is the same? Please explain your concerns.

File: plugins/TelescopeControl/src/ASCOM/TelescopeClientASCOM.cpp (hunk 3 above)
Comment (ShahriarNasseri) [reply plugins/TelescopeControl/src/ASCOM/TelescopeClientASCOM.cpp:31]: actually in the mathematical form which counted the line graph form, it reduces on the line methods on that mathematical form but on codding which you use the encapsulation the format of files which uses on graph methods has the problem on whole of the mathematical which changes the lines of flows as on physical and mathematical on both, the system file of boolian numbers intelligences as the real local form of lines and on the other hand the shape method has the static form as you know, but when it make on statistical form the base of the statics has the both problem as in this side StelUtils.hpp and StelTranslator.hpp too on the mirror side, i mean At the spase side of Coding we use the path directories but on the AI form we use the both of file formats as the space of directory and the definition of the both.

File: plugins/TelescopeControl/src/Lx200/TelescopeClientDirectLx200.hpp (hunk 2 above)
Comment (ShahriarNasseri) [reply plugins/TelescopeControl/src/Lx200/TelescopeClientDirectLx200.hpp:92]: the schematic of v and omega, on the mathematical form and in this connected planes interpolatedPosition as V which found on i module and sees iPdP coding measurements counted and omega as space after return base.

**Comment by ShahriarNasseri** [comment]
I taught we can solve it on writing the software but I'm working to see the result of the platform method for solving on the shape contrast on euclidian methods or lemmas.

**Comment by ShahriarNasseri** [comment]
V555 vector space is my the correct path but on connections

File: plugins/TelescopeControl/src/ASCOM/TelescopeClientASCOM.cpp (hunk 3 above)
Comment (gzotti) [reply plugins/TelescopeControl/src/ASCOM/TelescopeClientASCOM.cpp:31]: Seems you are wasting my time.

**Comment by kantn8r** [comment]
I'm not a programmer and do not really understand how this process works but I can test to confirm if my issue related to #2717 has been resolved. I just need to know when to re-install 1.22.3 again and test it.

**Comment by blakesphere** [comment]
hello, it seems there some effort being put into the Telescope Control plug-in. Excellent, as I'm awaiting the fix after 1.0.

I have a feature request, if it has not already been considered. In the slew-to window, I like all the controls, centre on object, centre on screen, the big Slew button. But I have always thought it very strange there's no big STOP button. When I teach people how to control their mount with Stellarium, I strongly advise they learn, memorise, and practice the abort slew keyboard shortcut to avoid a crash or snag. **_I point out it cannot be done in the GUI._** So, is it possible to add a Stop or Halt Slew button, please? There's space... ;-)

Lower priority, I wish the simulator would accept the Stop Slew keyboard shortcut. It does not. It would be good, I believe, if the simulator supported ALL the options available as with a real mount.

**Comment by gzotti** [comment]
@alex-w I think this is ready now. I'd like to do a "rebase and merge" this time, not squash.

**Comment by kantn8r** [comment]
I tried the weekly snapshot and Telescope Control seems to be working with Gemini Telescope.NET but when switching between Stellarium and SharpCap, Stellarium would close. I tried the same thing three times and each time when switching from Stellarium to SharpCap and the back to Stellarium it would close.

Windows 10 Pro, Celeron J4125 Mini-PC, SharpCap 4.0.9420 64-bit, Gemini Telescope.NET v1.0.84

**Comment by alex-w** [comment]
@kantn8r please attach log from Stellarium and from Telescope Control plugin

**Comment by kantn8r** [comment]
I'm attaching the Stellarium log. I'm not sure where to find the log from the Telescope Control plugin.
[strllarium_log_10232022.txt](https://github.com/Stellarium/stellarium/files/9855678/strllarium_log_10232022.txt)
 
If you tell me where to find the Telescope Plugin log I'll attach it.

**Comment by gzotti** [comment]
stellarium-047
See this setting. But @alex-w the location of these logs are indeed undocumented. Not sure where they can be found, I don't use this feature.

**Comment by kantn8r** [comment]
OK, I checked the log driver messages in Telescope Control and here is a new log.
[stellariun_log2_10232022.txt](https://github.com/Stellarium/stellarium/files/9855839/stellariun_log2_10232022.txt)

I can make Stellarium quit every time after a slew by switching from SharpCap to Stellarium.

**Comment by kantn8r** [comment]
I've reinstalled Stellarium 0.22.2 just to make sure it's not SharpCap and everything works fine.

**Comment by gzotti** [comment]
From the last-sent log I can only deduce that Stellarium stopped writing more, i.e. has crashed and did not shut down. 
It appears there should be files called "`log_TelescopeServer<n>.txt`" lext to log.txt that may include these additionally logged messages.

**Comment by blakesphere** [comment]
Thank you, Stellarium team! The Telescope Control plug-in is working great in 1.1. OK button non-response fixed. Very happy to see the Abort Slew button. Happy that abort works also in the simulator 'scope. Fantastic work!

**Comment by gzotti** [comment]
Actually the simulator is the only one that I can test. I hope instrumental aborts also work!

**Comment by blakesphere** [comment]
It works on a real mount! I tested 1.1 with my old Vixen Super Polaris which I have hacked with the IDEA GoToStar motor drive system. I use the ASCOM 32-bit driver for the Vixen SkySensor 2000 which gives near 100% compatibility. Again, the Abort Slew works! Same as me hitting the STOP button on the hand controller. Now I can halt the mount by remote control in the GUI (as well as the old keyboard shortcut). Thanks, again!

**Pull request merged**